sound/soc/sof/sof-audio.c

```diff
@@ -394,29 +396,56 @@ static int sof_free_widgets_in_path(struct snd_sof_dev *sdev, struct snd_soc_dap
  * The error path in this function ensures that all successfully set up widgets getting freed.
  */
 static int sof_set_up_widgets_in_path(struct snd_sof_dev *sdev, struct snd_soc_dapm_widget *widget,
-				      int dir, struct snd_soc_dapm_widget_list *list)
+				      int dir, struct snd_sof_pcm *spcm)
 {
+	struct snd_sof_pipeline_trigger_info *pipeline_info = &spcm->stream[dir].pipeline_info;
+	struct snd_soc_dapm_widget_list *list = spcm->stream[dir].list;
+	struct snd_sof_widget *swidget = widget->dobj.private;
+	struct snd_sof_widget *pipe_widget;
 	struct snd_soc_dapm_path *p;
 	int ret;
 
-	if (widget->dobj.private) {
+	if (swidget) {
+		int i;
+
 		ret = sof_widget_setup(sdev, widget->dobj.private);
 		if (ret < 0)
 			return ret;
+
+		/* skip populating the pipeline_list array if it is NULL */
+		if (!pipeline_info->pipeline_list)
+			goto sink_setup;
+
+		/*
+		 * Add the widget's pipe_widget to the list of pipelines to be triggered if not
+		 * already in the list. This will result in the pipelines getting added in the
+		 * order source to sink.
+		 */
```

Comment:
what happens if we have a split/merge structure, as done in some topologies. Is this ok if the intermediate pipelines sharing the same parent are exposed in no specific order?

Also what happens if we have a fork, e.g. if we play on two interfaces. Which sink should be triggered first?  would that even matter?

Reply:
> what happens if we have a split/merge structure, as done in some topologies. Is this ok if the intermediate pipelines sharing the same parent are exposed in no specific order?
> 
> Also what happens if we have a fork, e.g. if we play on two interfaces. Which sink should be triggered first? would that even matter?

I really don't think a fork is a special case here. If you consider only one side of the fork, the pipelines are triggered sink->source and that should be enough